packages/relayer/indexer/indexer.go

```diff
@@ -0,0 +1,383 @@
+
+	queue queue.Queue
+
+	srcChainId  *big.Int
+	destChainId *big.Int
+
+	watchMode WatchMode
+	syncMode  SyncMode
+
+	srv      *http.Server
+	httpPort uint64
+
+	wg *sync.WaitGroup
+}
+
+func (i *Indexer) InitFromCli(ctx context.Context, c *cli.Context) error {
+	cfg, err := NewConfigFromCliContext(c)
+	if err != nil {
+		return err
+	}
+
+	return InitFromConfig(ctx, i, cfg)
+}
+
+func InitFromConfig(ctx context.Context, i *Indexer, cfg *Config) (err error) {
+	db, err := cfg.OpenDBFunc()
+	if err != nil {
+		return err
+	}
+
+	eventRepository, err := repo.NewEventRepository(db)
+	if err != nil {
+		return err
+	}
+
+	blockRepository, err := repo.NewBlockRepository(db)
+	if err != nil {
+		return err
+	}
+
+	srcEthClient, err := ethclient.Dial(cfg.SrcRPCUrl)
+	if err != nil {
+		return err
+	}
+
+	destEthClient, err := ethclient.Dial(cfg.DestRPCUrl)
+	if err != nil {
+		return err
+	}
+
+	srv, err := http.NewServer(http.NewServerOpts{
+		EventRepo:     eventRepository,
+		Echo:          echo.New(),
+		CorsOrigins:   cfg.CORSOrigins,
+		SrcEthClient:  srcEthClient,
+		DestEthClient: destEthClient,
+		BlockRepo:     blockRepository,
+	})
+	if err != nil {
+		return err
+	}
+
+	q, err := cfg.OpenQueueFunc()
+	if err != nil {
+		return err
+	}
+
+	srcBridge, err := bridge.NewBridge(cfg.SrcBridgeAddress, srcEthClient)
+	if err != nil {
+		return errors.Wrap(err, "bridge.NewBridge")
+	}
+
+	destBridge, err := bridge.NewBridge(cfg.DestBridgeAddress, destEthClient)
+	if err != nil {
+		return errors.Wrap(err, "bridge.NewBridge")
+	}
+
+	var taikoL1 *taikol1.TaikoL1
+	if cfg.SrcTaikoAddress != ZeroAddress {
+		taikoL1, err = taikol1.NewTaikoL1(cfg.SrcTaikoAddress, srcEthClient)
+		if err != nil {
+			return errors.Wrap(err, "taikol1.NewTaikoL1")
+		}
+	}
+
+	srcChainID, err := srcEthClient.ChainID(context.Background())
+	if err != nil {
+		return errors.Wrap(err, "srcEthClient.ChainID")
+	}
+
+	destChainID, err := destEthClient.ChainID(context.Background())
+	if err != nil {
+		return errors.Wrap(err, "destEthClient.ChainID")
+	}
+
+	i.blockRepo = blockRepository
+	i.eventRepo = eventRepository
+	i.srcEthClient = srcEthClient
+
+	i.bridge = srcBridge
+	i.destBridge = destBridge
+	i.taikol1 = taikoL1
+
+	i.blockBatchSize = cfg.BlockBatchSize
+	i.numGoroutines = int(cfg.NumGoroutines)
+	i.subscriptionBackoff = time.Duration(cfg.SubscriptionBackoff) * time.Second
+
+	i.queue = q
+
+	i.srv = srv
+	i.httpPort = cfg.HTTPPort
+
+	i.srcChainId = srcChainID
+	i.destChainId = destChainID
+
+	i.syncMode = cfg.SyncMode
+	i.watchMode = cfg.WatchMode
+
+	i.wg = &sync.WaitGroup{}
+
+	return nil
+}
+
+func (i *Indexer) Name() string {
+	return "indexer"
+}
+
+// TODO
+func (i *Indexer) Close(ctx context.Context) {
+	if err := i.srv.Shutdown(ctx); err != nil {
+		slog.Error("srv shutdown", "error", err)
+	}
+
+	i.wg.Wait()
+}
+
+// nolint: funlen
+func (i *Indexer) Start() error {
+	go func() {
+		if err := i.srv.Start(fmt.Sprintf(":%v", i.httpPort)); err != nil {
```

Comment:
Seems in `echo`'s doc, they are checking `err != http.ErrServerClosed` instead of `err != nil`, shall we also do that here? https://echo.labstack.com/docs/start-server